src/Castle.Facilities.AspNetCore/CastleWindsorRegistrationExtensions.cs

```diff
@@ -29,44 +29,185 @@ namespace Castle.Facilities.AspNetCore
+		/// <param name="configure">Configuration options for controlling registration and lifestyles of controllers, tagHelpers and viewComponents</param>
+		public static void AddCastleWindsor(this IServiceCollection services, IWindsorContainer container, Action<CastleWindsorRegistrationOptions> configure = null)
 		{
-			services.AddRequestScopingMiddleware(container.BeginScope);
-			services.AddCustomControllerActivation(container.Resolve);
-			services.AddCustomTagHelperActivation(container.Resolve);
-			services.AddCustomViewComponentActivation(container.Resolve);
-			container.Kernel.Resolver.AddSubResolver(new LoggerDependencyResolver(services));
-			container.Kernel.Resolver.AddSubResolver(new FrameworkConfigurationDependencyResolver(services));
+			AddFrameworkResolversToWindsor(services, container);
+			InstallFrameworkIntegration(services, container);
+
+			var options = new CastleWindsorRegistrationOptions(container, services);
+			(configure ?? ((o) => { }))(options);
+
+			AddApplicationComponentsToWindsor(container, options);
 		}
 
+
 		/// <summary>
-		/// Use this to register all Controllers, ViewComponents and TagHelpers
+		/// For making types available to the IServiceCollection using 'late bound' factories resolved via Windsor. This makes things like the @Inject directive in Razor work.
 		/// </summary>
-		/// <typeparam name="TTypeAssembly">Type from assembly to scan and register ASP.NET Core components</typeparam>
-		/// <param name="app">Application builder retained as extension</param>
-		/// <param name="container">Windsor container to register framework types in</param>
-		public static void UseCastleWindsor<TTypeAssembly>(this IApplicationBuilder app, IWindsorContainer container)
+		/// <param name="registration">The component registration that gets copied across to the IServiceCollection</param>
+		/// <param name="serviceType">The type of the service</param>
+		/// <param name="container">Windsor container which 'activators'/'cross wired' components resolve from</param>
+		/// <param name="serviceCollection">ASP.NET Core service collection from Microsoft.Extensions.DependencyInjection</param>
+		/// <param name="lifestyleType">The lifestyle type that needs to be copied across, only Singleton, Scoped and Transient are supported</param>
+		public static ComponentRegistration<TService> CrossWired<TService>(this ComponentRegistration<TService> registration, Type serviceType, IWindsorContainer container, IServiceCollection serviceCollection, LifestyleType lifestyleType = LifestyleType.Scoped) where TService : class
 		{
-			var assembly = typeof(TTypeAssembly).Assembly;
-			UseCastleWindsor(container, assembly);
+			registration.LifeStyle.Is(lifestyleType);
+
+			if (lifestyleType == LifestyleType.Transient)
+			{
+				serviceCollection.AddTransient(serviceType, p =>
+				{
+					container.RequireScope();
+					return container.Resolve(serviceType);
+				});
+			}
+
+			if (lifestyleType == LifestyleType.Scoped)
```

Comment:
Probably should be an `else if`.

Reply:
Or do we throw are LifestyleNotSupportedException at https://github.com/fir3pho3nixx/Windsor/blob/aspnet-core-windsor-final/src/Castle.Facilities.AspNetCore/CastleWindsorRegistrationExtensions.cs#L186?

We need to figure out https://github.com/castleproject/Windsor/pull/394#discussion_r183031893 before we decide this.

Reply:
Yep, throw that exception in the else statement 👍.